File: voice_service.py
#!/usr/bin/env python3
"""
Nebula Voice Service - Local Speech-to-Text and Text-to-Speech
Provides voice input/output capabilities using:
- faster-whisper for STT (speech recognition)
- Piper TTS for TTS (speech synthesis)
"""

import os
import io
import wave
import logging
from pathlib import Path
from typing import Optional, BinaryIO

try:
    from faster_whisper import WhisperModel
except ImportError:
    WhisperModel = None

try:
    from piper import PiperVoice
except ImportError:
    PiperVoice = None

logger = logging.getLogger("NebulaVoice")

class VoiceService:
    """
    Local voice processing service with STT and TTS capabilities.
    All processing happens offline on the user's machine.
    """
    
    def __init__(self, model_dir: Path, stt_model: str = "base.en", tts_voice: str = "en_US-lessac-medium"):
        self.model_dir = model_dir
        self.model_dir.mkdir(exist_ok=True, parents=True)
        
        # STT Model (faster-whisper)
        self.stt_model_name = stt_model
        self.stt_model: Optional[WhisperModel] = None
        
        # TTS Model (Piper)
        self.tts_voice_name = tts_voice
        self.tts_model: Optional[PiperVoice] = None
        
        # Concurrency Lock (Windows/File Safety)
        import threading
        self._lock = threading.Lock()
        
        logger.info("[Voice] Service initialized (lazy loading enabled)")
    
    def load_stt_model(self):
        """Load faster-whisper model for speech recognition."""
        if self.stt_model is not None:
            return  # Already loaded
        
        if WhisperModel is None:
            raise RuntimeError("faster-whisper not installed. Run: pip install faster-whisper")
        
        logger.info(f"[STT] Loading {self.stt_model_name} model...")
        
        # Use CPU with 8-bit quantization for efficiency
        self.stt_model = WhisperModel(
            self.stt_model_name,
            device="cpu",
            compute_type="int8",
            download_root=str(self.model_dir / "whisper")
        )
        
        logger.info("[STT] Model loaded successfully")
    
    # --- Helper: TTS Model Auto-Downloader ---
    def ensure_piper_model(self):
        """Ensures the Piper model exists, downloading it if necessary."""
        voice_dir = self.model_dir / "piper"
        voice_dir.mkdir(exist_ok=True, parents=True)
        
        onnx_file = voice_dir / f"{self.tts_voice_name}.onnx"
        json_file = voice_dir / f"{self.tts_voice_name}.onnx.json"
        
        if onnx_file.exists() and json_file.exists():
            return onnx_file, json_file
            
        logger.info(f"[TTS] Downloading missing voice model: {self.tts_voice_name}")
        
        # Hardcoded mirror URL for known voices (Safe fallback)
        # Mirrors: https://github.com/rhasspy/piper/releases/tag/v0.0.2
        # Using a reliable HF mirror
        base_url = "https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/en/en_US/lessac/medium"
        if self.tts_voice_name != "en_US-lessac-medium":
             logger.warning(f"[TTS] Auto-download only supports 'en_US-lessac-medium'. Please check path manually for {self.tts_voice_name}")
             return onnx_file, json_file

        import urllib.request
        try:
            if not onnx_file.exists():
                logger.info("Downloading ONNX model...")
                urllib.request.urlretrieve(f"{base_url}/en_US-lessac-medium.onnx", onnx_file)
            
            if not json_file.exists():
                logger.info("Downloading Config JSON...")
                urllib.request.urlretrieve(f"{base_url}/en_US-lessac-medium.onnx.json", json_file)
                
            logger.info("[TTS] Download complete.")
        except Exception as e:
            logger.error(f"[TTS] Download failed: {e}")
            if onnx_file.exists(): onnx_file.unlink()
            if json_file.exists(): json_file.unlink()
            raise

        return onnx_file, json_file

    def load_tts_model(self):
        """Load Piper TTS model for speech synthesis."""
        if self.tts_model is not None:
            return  # Already loaded
        
        if PiperVoice is None:
            raise RuntimeError("piper-tts not installed. Run: pip install piper-tts")
        
        logger.info(f"[TTS] Loading {self.tts_voice_name} voice...")
        
        try:
            # Ensure model exists (Download if missing)
            voice_path, config_path = self.ensure_piper_model()
            
            self.tts_model = PiperVoice.load(str(voice_path), config_path=str(config_path))
            logger.info("[TTS] Voice loaded successfully")
        except Exception as e:
            logger.error(f"[TTS] Failed to load voice: {e}")
            self.tts_model = None
            raise

    def transcribe_audio(self, audio_file: BinaryIO, language: str = "en") -> dict:
        """
        Transcribe audio file to text using faster-whisper.
        Robustly handles generic audio inputs by converting to clean WAV first.
        """
        import subprocess

        # Ensure single-threaded access to model (Windows file locking fix)
        with self._lock:
            self.load_stt_model()
            
            # Paths
            raw_path = self.model_dir / "raw_audio_input.bin" # Unknown extension
            clean_path = self.model_dir / "clean_audio.wav"
            
            # Save raw input to disk
            with open(raw_path, "wb") as f:
                f.write(audio_file.read())
        
            try:
                # 1. Convert to standardized WAV using system FFmpeg
                # This fixes "Invalid Data" errors from WebM/Opus blobs or corrupt headers
                # We use raw_path input and force output to 16kHz mono PCM (ideal for Whisper)
                cmd = [
                    "ffmpeg", "-y",
                    "-i", str(raw_path),
                    "-ar", "16000",
                    "-ac", "1",
                    "-c:a", "pcm_s16le",
                    str(clean_path)
                ]
                
                # Run conversion, suppressing loud logging unless error
                result = subprocess.run(cmd, capture_output=True, text=True)
                
                if result.returncode != 0:
                    logger.error(f"[Voice] FFmpeg conversion failed: {result.stderr}")
                    # Fallback: Try raw path if ffmpeg failed (maybe it's already a valid wav?)
                    transcribe_target = str(raw_path)
                else:
                    transcribe_target = str(clean_path)

                # 2. Transcribe
                segments, info = self.stt_model.transcribe(
                    transcribe_target,
                    language=language if language != "auto" else None,
                    beam_size=5,
                    vad_filter=True
                )
                
                # Collect results
                text_parts = []
                segment_list = []
                
                for segment in segments:
                    text_parts.append(segment.text)
                    segment_list.append({
                        "start": segment.start,
                        "end": segment.end,
                        "text": segment.text
                    })
                
                full_text = " ".join(text_parts).strip()
                
                return {
                    "text": full_text,
                    "language": info.language,
                    "segments": segment_list,
                    "duration": info.duration
                }
            
            except Exception as e:
                logger.error(f"[Voice] Transcription critical failure: {e}")
                raise
            
            finally:
                # Cleanup
                try: raw_path.unlink(missing_ok=True) 
                except: pass
                try: clean_path.unlink(missing_ok=True) 
                except: pass
    
    def synthesize_speech(self, text: str, speed: float = 1.0) -> bytes:
        """
        Convert text to speech using Piper TTS.
        
        Args:
            text: Text to speak
            speed: Speech speed multiplier (1.0 = normal)
        
        Returns:
            WAV audio data as bytes
        """
        self.load_tts_model()
        
        if self.tts_model is None:
             raise RuntimeError("TTS model failed to load. Check logs.")

        # Generate audio
        audio_stream = io.BytesIO()
        
        try:
            # Piper generates raw PCM audio
            with wave.open(audio_stream, 'wb') as wav_file:
                wav_file.setnchannels(1)  # Mono
                wav_file.setsampwidth(2)  # 16-bit
                wav_file.setframerate(22050)  # Piper default sample rate
                
                # Synthesize (Piper outputs raw PCM)
                audio_bytes = self.tts_model.synthesize(text, length_scale=1.0/speed)
                wav_file.writeframes(audio_bytes)
            
            audio_stream.seek(0)
            return audio_stream.read()
        
        except Exception as e:
            logger.error(f"[TTS] Synthesis failed: {e}")
            raise

    # --- Streaming Support ---
    def create_stream_processor(self):
        """Create a new stream processor instance."""
        self.load_stt_model()
        return StreamProcessor(self.stt_model)

class StreamProcessor:
    """
    Handles streaming audio input with real-time incremental transcription.
    Uses sliding window with quick partial results.
    """
    def __init__(self, model: WhisperModel):
        self.model = model
        self.audio_buffer = bytearray()
        self.sample_rate = 16000
        self.sample_width = 2  # 16-bit
        
        # Process every 0.5 seconds for faster feedback
        self.process_interval = self.sample_rate * self.sample_width * 0.5
        
        # Keep last transcription to avoid duplication
        self.last_text = ""
        self.transcription_count = 0
        
    def add_audio(self, audio_chunk: bytes):
        """Add PCM audio chunk to buffer."""
        self.audio_buffer.extend(audio_chunk)
        
    def process(self) -> dict:
        """
        Process current buffer and return incremental text.
        Returns: { "text": "incremental text...", "is_final": False }
        """
        # Only process if we have enough audio (0.5s minimum)
        if len(self.audio_buffer) < self.process_interval:
            return None
        
        import numpy as np
        
        try:
            # Convert 16-bit PCM to float32 for Whisper
            audio_int16 = np.frombuffer(self.audio_buffer, dtype=np.int16)
            audio_float32 = audio_int16.astype(np.float32) / 32768.0
            
            # Quick transcription with low beam size for speed
            segments, _ = self.model.transcribe(
                audio_float32, 
                language="en", 
                beam_size=1,  # Faster than default 5
                word_timestamps=False,
                vad_filter=False  # Don't cut off speech aggressively
            )
            
            # Combine all segments
            text_parts = [segment.text for segment in segments]
            full_text = " ".join(text_parts).strip()
            
            # Only return if text changed
            if full_text != self.last_text:
                self.last_text = full_text
                self.transcription_count += 1
                
                return {
                    "text": full_text,
                    "is_final": False
                }
            
            return None
            
        except Exception as e:
            logger.error(f"[Stream] Processing failed: {e}")
            return None
            
    def finish(self) -> dict:
        """Finalize the stream and return final text."""
        if not self.audio_buffer:
            return {"text": self.last_text, "is_final": True}
        
        # Do one final transcription with higher quality
        import numpy as np
        
        try:
            audio_int16 = np.frombuffer(self.audio_buffer, dtype=np.int16)
            audio_float32 = audio_int16.astype(np.float32) / 32768.0
            
            # Final pass with better quality (beam_size=5)
            segments, _ = self.model.transcribe(
                audio_float32, 
                language="en", 
                beam_size=5,
                word_timestamps=False
            )
            
            text_parts = [segment.text for segment in segments]
            final_text = " ".join(text_parts).strip()
            
            # Clear state
            self.audio_buffer.clear()
            self.last_text = ""
            self.transcription_count = 0
            
            return {
                "text": final_text,
                "is_final": True
            }
            
        except Exception as e:
            logger.error(f"[Stream] Final processing failed: {e}")
            return {"text": self.last_text, "is_final": True}

# Global instance (lazy-loaded)
_voice_service: Optional[VoiceService] = None

def get_voice_service(model_dir: Path, stt_model: str = "base.en", tts_voice: str = "en_US-lessac-medium") -> VoiceService:
    """Get or create the global voice service instance."""
    global _voice_service
    if _voice_service is None:
        _voice_service = VoiceService(model_dir, stt_model, tts_voice)
    return _voice_service

if __name__ == "__main__":
    # Test the voice service
    import sys
    logging.basicConfig(level=logging.INFO)
    
    model_dir = Path(__file__).parent / "voice_models"
    service = VoiceService(model_dir)
    
    print("[Test] Loading STT model...")
    service.load_stt_model()
    print("[Test] STT model ready!")
    
    print("[Test] Loading TTS model...")
    service.load_tts_model()
    print("[Test] TTS model ready!")
    
    # Test TTS
    print("[Test] Generating speech...")
    test_text = "Hello! This is Nebula's local voice system. Everything runs on your computer, with complete privacy."
    audio_data = service.synthesize_speech(test_text)
    print(f"[Test] Generated {len(audio_data)} bytes of audio")
    
    # Save test audio
    test_file = model_dir / "test_output.wav"
    with open(test_file, "wb") as f:
        f.write(audio_data)
    print(f"[Test] Saved to: {test_file}")
    
    print("[Test] Voice service is working!")

File: rust_output/src/voice_service.rs
/// ZenAI Voice Service - Local Speech-to-Text and Text-to-Speech
/// Provides voice input/output capabilities using:
/// - faster-whisper for STT (speech recognition)
/// - Piper TTS for TTS (speech synthesis)

use anyhow::{Result, Context};
use crate::utils::{safe_print};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static _VOICE_SERVICE: std::sync::LazyLock<Option<VoiceService>> = std::sync::LazyLock::new(|| None);

/// Local voice processing service with STT and TTS capabilities.
/// All processing happens offline on the user's machine.
#[derive(Debug, Clone)]
pub struct VoiceService {
    pub model_dir: String,
    pub stt_model_name: String,
    pub stt_model: Option<WhisperModel>,
    pub tts_voice_name: String,
    pub tts_model: Option<PiperVoice>,
    pub _lock: std::sync::Mutex<()>,
}

impl VoiceService {
    /// Initialize instance.
    pub fn new(model_dir: PathBuf, stt_model: String, tts_voice: String) -> Self {
        Self {
            model_dir,
            stt_model_name: stt_model,
            stt_model,
            tts_voice_name: tts_voice,
            tts_model: None,
            _lock: std::sync::Mutex::new(()),
        }
    }
    /// Load faster-whisper model for speech recognition.
    pub fn load_stt_model(&mut self) -> Result<()> {
        // Load faster-whisper model for speech recognition.
        if self.stt_model.is_some() {
            return;
        }
        if WhisperModel.is_none() {
            return Err(anyhow::anyhow!("RuntimeError('faster-whisper not installed. Run: pip install faster-whisper')"));
        }
        logger.info(format!("[STT] Loading {} model...", self.stt_model_name));
        self.stt_model = WhisperModel(self.stt_model_name, /* device= */ "cpu".to_string(), /* compute_type= */ "int8".to_string(), /* download_root= */ (self.model_dir / "whisper".to_string()).to_string());
        Ok(logger.info("[STT] Model loaded successfully".to_string()))
    }
    /// Ensures the Piper model exists, downloading it if necessary.
    pub fn ensure_piper_model(&mut self) -> Result<()> {
        // Ensures the Piper model exists, downloading it if necessary.
        let mut voice_dir = (self.model_dir / "piper".to_string());
        voice_dir.create_dir_all();
        let mut onnx_file = (voice_dir / format!("{}.onnx", self.tts_voice_name));
        let mut json_file = (voice_dir / format!("{}.onnx.json", self.tts_voice_name));
        if (onnx_file.exists() && json_file.exists()) {
            (onnx_file, json_file)
        }
        logger.info(format!("[TTS] Downloading missing voice model: {}", self.tts_voice_name));
        let mut base_url = "https://huggingface.co/rhasspy/piper-voices/resolve/v1.0.0/en/en_US/lessac/medium".to_string();
        if self.tts_voice_name != "en_US-lessac-medium".to_string() {
            logger.warning(format!("[TTS] Auto-download only supports 'en_US-lessac-medium'. Please check path manually for {}", self.tts_voice_name));
            (onnx_file, json_file)
        }
        // TODO: import urllib::request
        // try:
        {
            if !onnx_file.exists() {
                logger.info("Downloading ONNX model...".to_string());
                urllib::request.urlretrieve(format!("{}/en_US-lessac-medium.onnx", base_url), onnx_file);
            }
            if !json_file.exists() {
                logger.info("Downloading Config JSON...".to_string());
                urllib::request.urlretrieve(format!("{}/en_US-lessac-medium.onnx.json", base_url), json_file);
            }
            logger.info("[TTS] Download complete.".to_string());
        }
        // except Exception as e:
        Ok((onnx_file, json_file))
    }
    /// Load Piper TTS model for speech synthesis.
    pub fn load_tts_model(&mut self) -> Result<()> {
        // Load Piper TTS model for speech synthesis.
        if self.tts_model.is_some() {
            return;
        }
        if PiperVoice.is_none() {
            return Err(anyhow::anyhow!("RuntimeError('piper-tts not installed. Run: pip install piper-tts')"));
        }
        logger.info(format!("[TTS] Loading {} voice...", self.tts_voice_name));
        // try:
        {
            let (mut voice_path, mut config_path) = self.ensure_piper_model();
            self.tts_model = PiperVoice.load(voice_path.to_string(), /* config_path= */ config_path.to_string());
            logger.info("[TTS] Voice loaded successfully".to_string());
        }
        // except Exception as e:
    }
    /// Transcribe audio file to text using faster-whisper.
    /// Robustly handles generic audio inputs by converting to clean WAV first.
    pub fn transcribe_audio(&mut self, audio_file: BinaryIO, language: String) -> Result<HashMap> {
        // Transcribe audio file to text using faster-whisper.
        // Robustly handles generic audio inputs by converting to clean WAV first.
        // TODO: import subprocess
        let _ctx = self._lock;
        {
            self.load_stt_model();
            let mut raw_path = (self.model_dir / "raw_audio_input.bin".to_string());
            let mut clean_path = (self.model_dir / "clean_audio.wav".to_string());
            let mut f = File::open(raw_path)?;
            {
                f.write(audio_file.read());
            }
            // try:
            {
                let mut cmd = vec!["ffmpeg".to_string(), "-y".to_string(), "-i".to_string(), raw_path.to_string(), "-ar".to_string(), "16000".to_string(), "-ac".to_string(), "1".to_string(), "-c:a".to_string(), "pcm_s16le".to_string(), clean_path.to_string()];
                let mut result = std::process::Command::new("sh").arg("-c").arg(cmd, /* capture_output= */ true, /* text= */ true, /* shell= */ false).output().unwrap();
                if result.returncode != 0 {
                    logger.error(format!("[Voice] FFmpeg conversion failed: {}", result.stderr));
                    let mut transcribe_target = raw_path.to_string();
                } else {
                    let mut transcribe_target = clean_path.to_string();
                }
                let (mut segments, mut info) = self.stt_model.transcribe(transcribe_target, /* language= */ if language != "auto".to_string() { language } else { None }, /* beam_size= */ 5, /* vad_filter= */ true);
                let mut text_parts = vec![];
                let mut segment_list = vec![];
                for segment in segments.iter() {
                    text_parts.push(segment.text);
                    segment_list.push(HashMap::from([("start".to_string(), segment.start), ("end".to_string(), segment.end), ("text".to_string(), segment.text)]));
                }
                let mut full_text = text_parts.join(&" ".to_string()).trim().to_string();
                HashMap::from([("text".to_string(), full_text), ("language".to_string(), info.language), ("segments".to_string(), segment_list), ("duration".to_string(), info.duration)])
            }
            // except Exception as e:
            // finally:
                // try:
                {
                    raw_path.unlink(/* missing_ok= */ true);
                }
                // except Exception as _e:
                // try:
                {
                    clean_path.unlink(/* missing_ok= */ true);
                }
                // except Exception as _e:
        }
    }
    /// Convert text to speech using Piper TTS.
    /// Optimized to use float32 PCM directly (no int16 conversion overhead).
    /// 
    /// Args:
    /// text: Text to speak
    /// speed: Speech speed multiplier (1.0 = normal)
    /// 
    /// Returns:
    /// WAV audio data as bytes (float32 PCM format for faster synthesis)
    pub fn synthesize_speech(&mut self, text: String, speed: f64) -> Result<Vec<u8>> {
        // Convert text to speech using Piper TTS.
        // Optimized to use float32 PCM directly (no int16 conversion overhead).
        // 
        // Args:
        // text: Text to speak
        // speed: Speech speed multiplier (1.0 = normal)
        // 
        // Returns:
        // WAV audio data as bytes (float32 PCM format for faster synthesis)
        self.load_tts_model();
        if self.tts_model.is_none() {
            return Err(anyhow::anyhow!("RuntimeError('TTS model failed to load. Check logs.')"));
        }
        let mut audio_stream = io.BytesIO();
        // try:
        {
            let mut audio_chunks = vec![];
            let mut sample_rate = 22050;
            for chunk in self.tts_model.synthesize(text).iter() {
                if /* hasattr(chunk, "audio_float_array".to_string()) */ true {
                    audio_chunks.push(chunk.audio_float_array.tobytes());
                    if /* hasattr(chunk, "sample_rate".to_string()) */ true {
                        let mut sample_rate = chunk.sample_rate;
                    }
                }
            }
            if !audio_chunks {
                return Err(anyhow::anyhow!("RuntimeError('TTS synthesize produced no audio chunks')"));
            }
            let mut pcm_data = audio_chunks.join(&b"");
            let mut wav_file = wave.open(audio_stream, "wb".to_string());
            {
                wav_file.setnchannels(1);
                wav_file.setsampwidth(4);
                wav_file.setframerate(sample_rate);
                wav_file.writeframes(pcm_data);
            }
            audio_stream.seek(0);
            audio_stream.read()
        }
        // except Exception as e:
    }
    /// Create a new stream processor instance.
    pub fn create_stream_processor(&self) -> () {
        // Create a new stream processor instance.
        self.load_stt_model();
        StreamProcessor(self.stt_model)
    }
}

/// Handles streaming audio input with real-time incremental transcription.
/// Uses sliding window with quick partial results.
#[derive(Debug, Clone)]
pub struct StreamProcessor {
    pub model: String,
    pub audio_buffer: bytearray,
    pub sample_rate: i64,
    pub sample_width: i64,
    pub process_interval: String,
    pub last_text: String,
    pub transcription_count: i64,
}

impl StreamProcessor {
    /// Initialize instance.
    pub fn new(model: WhisperModel) -> Self {
        Self {
            model,
            audio_buffer: bytearray(),
            sample_rate: 16000,
            sample_width: 2,
            process_interval: ((self.sample_rate * self.sample_width) * 0.5_f64),
            last_text: "".to_string(),
            transcription_count: 0,
        }
    }
    /// Add PCM audio chunk to buffer.
    pub fn add_audio(&self, audio_chunk: Vec<u8>) -> () {
        // Add PCM audio chunk to buffer.
        self.audio_buffer.extend(audio_chunk);
    }
    /// Process current buffer and return incremental text.
    /// Returns: { "text": "incremental text...", "is_final": false }
    pub fn process(&mut self) -> Result<HashMap> {
        // Process current buffer and return incremental text.
        // Returns: { "text": "incremental text...", "is_final": false }
        if self.audio_buffer.len() < self.process_interval {
            None
        }
        // TODO: import numpy as np
        // try:
        {
            let mut audio_int16 = np.frombuffer(self.audio_buffer, /* dtype= */ np.int16);
            let mut audio_float32 = (audio_int16.astype(np.float32) / 32768.0_f64);
            let (mut segments, _) = self.model.transcribe(audio_float32, /* language= */ "en".to_string(), /* beam_size= */ 1, /* word_timestamps= */ false, /* vad_filter= */ false);
            let mut text_parts = segments.iter().map(|segment| segment.text).collect::<Vec<_>>();
            let mut full_text = text_parts.join(&" ".to_string()).trim().to_string();
            if full_text != self.last_text {
                self.last_text = full_text;
                self.transcription_count += 1;
                HashMap::from([("text".to_string(), full_text), ("is_final".to_string(), false)])
            }
            None
        }
        // except Exception as e:
    }
    /// Finalize the stream and return final text.
    pub fn finish(&mut self) -> Result<HashMap> {
        // Finalize the stream and return final text.
        if !self.audio_buffer {
            HashMap::from([("text".to_string(), self.last_text), ("is_final".to_string(), true)])
        }
        // TODO: import numpy as np
        // try:
        {
            let mut audio_int16 = np.frombuffer(self.audio_buffer, /* dtype= */ np.int16);
            let mut audio_float32 = (audio_int16.astype(np.float32) / 32768.0_f64);
            let (mut segments, _) = self.model.transcribe(audio_float32, /* language= */ "en".to_string(), /* beam_size= */ 5, /* word_timestamps= */ false);
            let mut text_parts = segments.iter().map(|segment| segment.text).collect::<Vec<_>>();
            let mut final_text = text_parts.join(&" ".to_string()).trim().to_string();
            self.audio_buffer.clear();
            self.last_text = "".to_string();
            self.transcription_count = 0;
            HashMap::from([("text".to_string(), final_text), ("is_final".to_string(), true)])
        }
        // except Exception as e:
    }
}

/// Get or create the global voice service instance.
pub fn get_voice_service(model_dir: PathBuf, stt_model: String, tts_voice: String) -> VoiceService {
    // Get or create the global voice service instance.
    // global/nonlocal _voice_service
    if _voice_service.is_none() {
        let mut _voice_service = VoiceService(model_dir, stt_model, tts_voice);
    }
    _voice_service
}
